text/2017-08-30-simplify-defer.md

```diff
@@ -0,0 +1,182 @@
+    Ok(()) => unsafe {
+        scope.defer_drop(head);
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+Neither `Atomic` nor `ManuallyDrop` have a destructor, so dropping a `Node` will be a noop,
+which makes `defer_drop` in this case equivalent to `defer_free`.
+This makes `defer_free` unnecessary.
+This is the kind of problem `ManuallyDrop` was created to solve.
+
+### Do we need `defer_free`?
+
+Let's take a step further and rewrite the same piece of code using `defer`:
+
+```rust
+match self.head.compare_and_set_weak(head, next, AcqRel, scope) {
+    Ok(()) => unsafe {
+        let raw = head.as_raw() as usize;
+        scope.defer(|| drop(Box::from_raw(raw as *const Node<T>)));
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+Calling the destructor will now come with some performance cost due to additional boxing
+(`defer` stores `FnOnce()` in a `Box`).
+Moreover, deferring destruction this way is a bit more unergonomic.
+Let's fix both problems...
+
+# Detailed design
+
+### `Deferred`
+
+Instead of having the triplet `defer_free`/`defer_drop`/`defer`, we'll have only `defer`
+with the same interface, but this time as a safe function:
+
+```rust
+impl Scope {
+    // Deferred execution of arbitrary function `f`.
+    fn defer<F: FnOnce() + Send + 'static>(&self, f: F);
+
+    // ...
+}
+```
+
+In most cases `defer` will be used to just deallocate memory or execute simple destruction
+routines. The passed `FnOnce()` closure will almost always fit into, say, 3 words.
+
+I have an [implementation](https://gist.github.com/stjepang/00d63b8febc07b297eae3480e80d0e91)
+of `Deferred` ready, which is able to store a small closure within itself, falling back
+to heap allocation for big closures.
+
+**TL;DR:** `Deferred` is to `FnOnce() + Send + 'static` what `SmallVec` is to `Vec`.
+
+### `Owned::from_ptr`
+
+Finally, let's reconsider the piece of code from the motivation section and make it more ergonomic.
+There are several unergonomic obstacles to destruction of `head`:
+
+1. A `Ptr<'scope>` cannot be passed to a `'static` closure.
+2. A raw pointer cannot be passed to a `Send` closure (a workaround is to convert the pointer to `usize`).
+3. Finally, the raw pointer must be passed to `Box::from_raw`.
+
+To simplify matters, let's introduce a new unsafe constructor, `Owned::from_ptr`:
+
+```rust
+match self.head.compare_and_set_weak(head, next, AcqRel, scope) {
+    Ok(()) => unsafe {
+        let head = Owned::from_ptr(head);
+        scope.defer(move || drop(head));
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+Strictly speaking, this code is possibly incorrect because an `Owned` and `Ptr` pointing to the
+same thing exist at the same time. This is okay because the `Owned` will not be used until the
+closure executes, but **@RalfJung**'s new unsafe code checker might reject the code as unsound.
```

Comment:
I feel the proposed solution is hacky for the reason that the meaning of `Owned` or `Ptr<'static>` is unclear.  I think it would be great if the meaning can be explained in [RustBelt](http://www.mpi-sws.org/~dreyer/papers/rustbelt/paper.pdf)'s lifetime logic.

Also, I doubt that the `defer*` family of functions is safe.  For example, the client should guarantee that the `defer_drop()`ed object is no longer pointed to by a memory cell.  This knowledge cannot be expressed in Rust's type system, so it should be `unsafe`.

Reply:
Regarding conversion from `Ptr<'scope>` to `Ptr<'static>`, that is definitely hacky. But I think what we *actually* want here is conversion to `Ptr<'unsafe>`. Unsafe lifetime was proposed in [this RFC](https://github.com/rust-lang/rfcs/pull/1918), but in the end it was postponed.

Regarding safety, `defer_free` and `defer_drop` are definitely unsafe and this RFC doesn't aim to change that. However, `defer` is currently unsafe, but this RFC proposes to make it a safe function. I suppose you're okay with that?

Reply:
Ah, yes.  I am okay with `pub safe fn defer()`.